### PR TITLE
Add disallowed_headers knob to ext_authz config

### PR DESCRIPTION
**Pull request by antoniovleonti**
Commit Message: Add disallowed_headers knob to ext_authz config
Additional Description: This PR adds a config field to ext_authz that will prevent the specified headers from being sent to the external authentication service. This new field (disallowed_headers) will always override the allowed_headers field.
Risk Level: low
Testing: unit tests & integration tests
Docs Changes: none
Release Notes: added to changelog
Platform Specific Features: n/a
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by tyxia** [comment]
/assign @yanjunxiang-google 

Could you take a pass on it? Thanks

**Comment by antoniovleonti** [comment]
Just realized, #33244 should probably be merged before this one.

#33244 refactors HttpFilterTestParam, which is a parameterized test fixture used to write some of this PR's tests.

**Review by wbpcode — COMMENTED** [review]
/lgtm api

**Comment by wbpcode** [comment]
ping me if this is ready for a maintainer review. :)

**Review by wbpcode — COMMENTED** [review]
This looks good to me overall. But do you get agreement on this feature internally? @yanjunxiang-google

**Comment by antoniovleonti** [comment]
@yanjunxiang-google please review

File: source/extensions/filters/common/ext_authz/check_request_utils.cc
Comment (tyxia) [inline source/extensions/filters/common/ext_authz/check_request_utils.cc:160]: it seems that we can do following to emphasize that `disallowed_headers` take precedence?

```
if (disallowed_headers_matcher != nullptr && disallowed_headers_matcher->matches(key)) {
  continue;
}

if (allowed_headers_matcher != nullptr && !allowed_headers_matcher->matches(key)) {
      return Continue;
 }
```

**Review by tyxia — COMMENTED** [review]
LGTM modulo one comment above.

Thank you for the work!

File: source/extensions/filters/common/ext_authz/check_request_utils.cc
Comment (antoniovleonti) [reply source/extensions/filters/common/ext_authz/check_request_utils.cc]: Agreed, done.

**Comment by antoniovleonti** [comment]
@wbpcode I think I need your API review again.

**Review by tyxia — APPROVED** [review]

**Comment by ravenblackx** [comment]
@wbpcode ping for api re-review

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks for your contribution. And sorry for the delay, I am a little busy recently and didn't respond to PR reviews in timely.

**Comment by wbpcode** [comment]
Because @tyxia is the owner and has approved this PR. And it's LGTM, so will merge it directly. :)

**Pull request merged**
